src/engine/Server.cpp

```diff
@@ -793,9 +793,15 @@ boost::asio::awaitable<void> Server::processQuery(
 template <std::invocable Function, typename T>
 Awaitable<T> Server::computeInNewThread(Function function,
                                         SharedCancellationHandle handle) {
-  auto timerRunning = std::make_shared<std::atomic_flag>(true);
-  auto inner = [function = std::move(function), timerRunning]() mutable -> T {
-    timerRunning->clear();
+  std::promise<std::function<void()>> callbackPromise{};
+  auto callbackFuture = callbackPromise.get_future();
+
+  auto inner = [function = std::move(function),
+                callbackFuture = std::move(callbackFuture)]() mutable -> T {
+    // Ensure future is ready by the time this is called.
+    AD_CORRECTNESS_CHECK(callbackFuture.wait_for(std::chrono::milliseconds{
+                             0}) == std::future_status::ready);
+    callbackFuture.get()();
```

Comment:
give the `callbackFuture` a better name, and shortly describe here what it does (
We want timeout checks while we are waiting for a thread, but after that we don't need them anymore).